frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx

```diff
@@ -61,255 +64,242 @@ import { formatNumberToLength, formatLengthAsVx } from "libs/format_utils";
 import { api } from "oxalis/singletons";
 import { ChangeColorMenuItemContent } from "components/color_picker";
 import { HideTreeEdgesIcon } from "./hide_tree_eges_icon";
+import { useEffectOnlyOnce } from "libs/react_hooks";
+import { ColoredDotIcon } from "./segments_tab/segment_list_item";
 
-const CHECKBOX_STYLE = { marginLeft: 4 };
-const CHECKBOX_PLACEHOLDER_STYLE = {
-  width: 16,
-  display: "inline-block",
-};
-type OwnProps = {
+type Props = {
   activeTreeId: number | null | undefined;
   activeGroupId: number | null | undefined;
-  treeGroups: Array<TreeGroup>;
+  treeGroups: TreeGroup[];
   sortBy: string;
   trees: TreeMap;
-  selectedTrees: Array<number>;
-  onSelectTree: (arg0: number) => void;
+  selectedTreeIds: number[];
+  onSingleSelectTree: (arg0: number) => void;
+  onMultiSelectTree: (arg0: number) => void;
   deselectAllTrees: () => void;
   onDeleteGroup: (arg0: number) => void;
   allowUpdate: boolean;
 };
-type Props = OwnProps & {
-  onShuffleTreeColor: (arg0: number) => void;
-  onSetActiveTree: (arg0: number) => void;
-  onSetActiveTreeGroup: (arg0: number) => void;
-  onToggleTree: (arg0: number) => void;
-  onDeleteTree: (arg0: number) => void;
-  onToggleAllTrees: () => void;
-  onSetTreeColor: (arg0: number, arg1: Vector3) => void;
-  onToggleTreeGroup: (arg0: number) => void;
-  onUpdateTreeGroups: (arg0: Array<TreeGroup>) => void;
-  onBatchActions: (arg0: Array<Action>, arg1: string) => void;
-  onToggleHideInactiveTrees: () => void;
-  onShuffleAllTreeColors: () => void;
-  onSetTreeEdgesVisibility: (treeId: number, edgesAreVisible: boolean) => void;
-  onSetTreeType: (treeId: number, type: TreeTypeEnum) => void;
-};
-type State = {
-  prevProps: Props | null | undefined;
-  expandedGroupIds: Record<number, boolean>;
-  groupTree: Array<TreeNode>;
-  searchFocusOffset: number;
-  activeTreeDropdownId: number | null | undefined;
-  activeGroupDropdownId: number | null | undefined;
-};
 
 export type GenerateNodePropsType = {
   title?: JSX.Element;
   className?: string;
   style?: React.CSSProperties;
 };
 
-const didTreeDataChange = (prevProps: Props, nextProps: Props): boolean =>
-  prevProps.trees !== nextProps.trees ||
-  prevProps.treeGroups !== nextProps.treeGroups ||
-  prevProps.sortBy !== nextProps.sortBy;
-
-class TreeHierarchyView extends React.PureComponent<Props, State> {
-  state: State = {
-    expandedGroupIds: {
-      [MISSING_GROUP_ID]: true,
-    },
-    groupTree: [],
-    prevProps: null,
-    searchFocusOffset: 0,
-    activeTreeDropdownId: null,
-    activeGroupDropdownId: null,
-  };
-
-  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
-    if (prevState.prevProps == null || didTreeDataChange(prevState.prevProps, nextProps)) {
```

Comment:
`didTreeDataChange` change has been replaces with `React.memo(<TreeHierarchyView/>)` at the very bottom of this file.